### PR TITLE
added-checkbox-style

### LINKED ISSUE #109: Add Checkbox Style to InputField
x-team/auto#748 uses a checkbox input. If you try to use a checkbox with InputField its renders like
captura de pantalla 2018-01-25 a la s 11 08 02 p m

### PR DESCRIPTION
**Pull request by jepz20**
Resolves #109 

checkbox

### CODE REVIEW

File: src/components/forms/InputField.js [hunk 1]
```diff
@@ -57,23 +57,23 @@ const radioInputStyles = {
...
+)
+
 const ComponentRoot = elem.div()
 const FieldRoot = elem.div(cmz(`
   display: inline-block
   position: relative
 `))
 const RadioCircle = elem.span(cmz(radioInputStyles.circle))
-const RadioLabel = elem.label(cmz(radioInputStyles.label))
+const RadioLabel = elem.label(cmz(label))
```
Comment (falcoagustin) [inline src/components/forms/InputField.js]: Label is already a cmz, it does not need to be cmz'ed again 🙂

File: src/components/forms/InputField.js [hunk 2]
```diff
@@ -107,6 +107,51 @@ const errorInput = cmz(`
...
+      border-top: none
+      border-right: none
+      transform: rotate(-45deg)
+      transition: all 0.25s ease
+    }
+  `)
+}
+
+const CheckboxTick = elem.span(checkboxInputStyles.tick)
+const Checkboxlabel = elem.label(cmz(label))
```
Comment (falcoagustin) [inline src/components/forms/InputField.js]: same here

File: src/components/forms/InputField.js [hunk 3]
```diff
@@ -57,23 +57,23 @@ const radioInputStyles = {
...
-  label: cmz(
-    typo.baseText,
-    `
-      margin-left: 30px
-    `
-  )
+  `)
 }
 
+const label = cmz(
```
Comment (falcoagustin) [inline src/components/forms/InputField.js]: We could rename this variable labelStyles, what do you think ?

Comment (jepz20) [reply src/components/forms/InputField.js]: make sense, thanks 😄

**Comment by jepz20** [comment]
Hey @falcoagustin thanks for the review, I have updated the requested changes, could you check it again, please 😄

File: src/components/forms/InputField.js [hunk 4]
```diff
@@ -107,6 +107,51 @@ const errorInput = cmz(`
...
+    }
+
+    &:after {
+      opacity: 0
+      content: ' '
+      position: absolute
+      width: 8px
+      height: 4px
+      top: 5px
+      left: 5px
```
Comment (falcoagustin) [inline src/components/forms/InputField.js]: Can you change this to 4px? See this difference, I know its minimal, but I looked at it at the very first try,
screen shot 2018-01-29 at 4 34 32 pm
vs
screen shot 2018-01-29 at 4 34 36 pm
Can you spot the difference?

File: src/components/forms/InputField.js [hunk 5]
```diff
@@ -170,15 +227,15 @@ class InputField extends PureComponent<Props> {
 
     const inputId = id || name
     const labelId = inputId ? `label-${inputId}` : ''
-    const isRadio = type === 'radio'
-    const RootComponent = isRadio ? FieldRoot : ComponentRoot
+    const isRadioOrCheckbox = (type === 'radio' || type === 'checkbox')
```
Comment (falcoagustin) [inline src/components/forms/InputField.js]: We use this already in two places, what do you think about extracting this into a common function? This way you can use it when checking what type to render and underneath,
```js
const RootComponent = isRadioOrCheckbox() ? FieldRoot : ComponentRoot
```

**Comment by nicksp** [comment]
Guys, sorry to disturb you. I'm late to the party and do not want to break things... but have you checked I've already got `InputField` component that supports textareas, regular inputs, radio buttons? It can easily be extended to support checkboxes. This way we'll have a generic component that can handle all the required logic in a common manner, consistently and with no duplication. Wdyt? 😄

**Comment by nicksp** [comment]
Ignore my prev. comment. Late time here 😆

File: src/components/forms/InputField.js (hunk 5 above)
Comment (jepz20) [reply src/components/forms/InputField.js]: Thanks for the comment @falcoagustin, this actually made me rethink the name to a more generic one. So I change it to `isSpecialType` and is based on the `specialTypesDefinitions` object.

File: src/components/forms/InputField.js (hunk 4 above)
Comment (jepz20) [reply src/components/forms/InputField.js]: Nice catch, thanks 😄

**Comment by jepz20** [comment]
Thanks again, @falcoagustin mind to give it another go 🙂

File: src/components/forms/InputField.js [hunk 6]
```diff
@@ -107,10 +107,70 @@ const errorInput = cmz(`
...
+}
+
+const CheckboxTick = elem.span(checkboxInputStyles.tick)
+const Checkboxlabel = elem.label(labelStyles)
+
 const getTagName = type => type === 'textarea' ? 'textarea' : 'input'
 
 const inputFactory = type => elem[getTagName(type)](inputStyles)
 
+const specialTypesDefinitions : Object = {
```
Comment (jepz20) [inline src/components/forms/InputField.js]: There is an issue with flow where if not setting specialTypesDefinitions to `Object` it throws an error that types doesn't exist even though it is already being validated in isSpecialType base on [this answer](facebook/flow#1606 (comment)), I set the specialTypesDefinitions type as object (an unsealed object) to fix the issue

**Comment by piotrpasich** [comment]
What's the status of this PR ? What do we need to do to make it ready ?

**Comment by nicksp** [comment]
will give it a check later today 😄

**Comment by falcoagustin** [comment]
Looks good to go from my side 😄

File: src/components/forms/InputField.js [hunk 7]
```diff
@@ -57,23 +57,23 @@ const radioInputStyles = {
...
+    margin-left: 30px
+  `
+)
+
 const ComponentRoot = elem.div()
 const FieldRoot = elem.div(cmz(`
   display: inline-block
   position: relative
 `))
 const RadioCircle = elem.span(cmz(radioInputStyles.circle))
```
Comment (nicksp) [inline src/components/forms/InputField.js]: actually `radioInputStyles.circle` is also wrapped with `cmz` already. You can remove double wrapper from this line 😄

File: src/components/forms/InputField.js [hunk 8]
```diff
@@ -107,10 +107,70 @@ const errorInput = cmz(`
...
+      border-top: none
+      border-right: none
+      transform: rotate(-45deg)
+      transition: all 0.25s ease
+    }
+  `)
+}
+
+const CheckboxTick = elem.span(checkboxInputStyles.tick)
+const Checkboxlabel = elem.label(labelStyles)
```
Comment (nicksp) [inline src/components/forms/InputField.js]: this is similar to the prev. definition `const RadioLabel = elem.label(labelStyles)`. We can define generic label definition once to remove duplication, right 😃 ? Smth. like:

```js
const Label = elem.label(...)
```

File: src/components/forms/InputField.js [hunk 9]
```diff
@@ -134,11 +194,11 @@ class InputField extends PureComponent<Props> {
     const labelId = inputId ? `label-${inputId}` : ''
     const errorClassName = isInvalid ? errorInput : ''
 
-    if (type === 'radio') {
+    if (isSpecialType(type)) {
```
Comment (nicksp) [inline src/components/forms/InputField.js]: Let's do object destructuring after this line, to pull props off the object. This way we can condense the below render code to this:

```js
const { className, box: ElemBox, label: ElemLabel } = specialTypesDefinitions[type]
<...>
Tag({
  className,
  type,
  <...>
}),
ElemBox(),
ElemLabel(label)
```

Wdyt?

File: src/components/forms/InputField.js (hunk 7 above)
Comment (jepz20) [reply src/components/forms/InputField.js]: nice catch thanks 😄

File: src/components/forms/InputField.js (hunk 8 above)
Comment (jepz20) [reply src/components/forms/InputField.js]: you are right, makes sense. Thanks

File: src/components/forms/InputField.js (hunk 9 above)
Comment (jepz20) [reply src/components/forms/InputField.js]: this does look nicer, thanks for the recommendation 😄

**Comment by jepz20** [comment]
Hey @nicksp thanks for the review 😄 .

Ready for another re-review, I did your recommendation to have the whole label clickable, since I think it indeed provides a better experience for the user.

**Comment by nicksp** [comment]
Updated snapshots and we're ready for merge 😄

**Pull request merged**